_posts/2018-10-09-enabling-ecc-on-pc-engines-apu2.md

```diff
@@ -0,0 +1,371 @@
+flipped bits scrubbing has to be done periodically, before multiple errors
+within one ECC word can occur - time between scrubs isn't fixed, and according
+to [BKDG, 52740 Rev 3.06](https://www.amd.com/system/files/TechDocs/52740_16h_Models_30h-3Fh_BKDG.pdf):
+
+> There are many factors which influence scrub rates. Among these are:
+> * The size of memory or cache to be scrubbed
+> * Resistance to upsets
+> * Geographic location and altitude
+> * Alpha particle contribution of packaging
+> * Performance sensitivity
+> * Risk aversion
+
+Usually, ECC can fix only single bit, and detect two changed bits in ECC word,
+but with AMD Embedded G series GX-412TC SoC included in PC Engines apu2 platform
+a variation named `x4 ECC` is used:
+
+> The x4 code uses thirty-six 4-bit symbols to make a 144-bit ECC word made up
+> of 128 data bits and 16 check bits. The x4 code is a single symbol correcting
+> (SSC) and a double symbol detecting (DSD) code. This means the x4 code is able
+> to correct 100% of single symbol errors (any bit error combination within one
+> symbol), and detect 100% of double symbol errors (any bit error combination
+> within two symbols).
+
+ECC feature requires support from the memory controller (on SoC), memory banks,
+connections between them on a motherboard (memory bus is wider to support
+reading check bits at the same time as main data) and firmware which sets
+everything up. Further operation is transparent to operating systems, but it is
+possible for them to gather information about how often errors are being
+corrected.
+
+It is possible to turn off ECC check and error reporting for one range of
+memory. This range is used as a framebuffer for integrated GPU - scrubbing could
+have an impact on framerate and things like V-sync, and one changed pixel
+visible only for a fraction of second is acceptable.
+
+## Testing for ECC support
+
+To be sure that ECC works one must notice a corrected ECC error. There are few
+problems to this:
+* ECC has to be available and properly configured
+* ECC error reporting must be enabled
+* there must be a correctable ECC error
+
+First two points are something that can and has to be done in firmware as long
+as a hardware is compatible, so enabling it takes some work and a whole lot of
+research, but is definitely possible.
+
+The last issue is actually the hardest one. It is impossible to reliably
+simulate e.g. a cosmic ray, and row hammering takes a lot of time and requires
+specific knowledge of the internal structure of a memory bank. To help with
+testing vendors of memory controllers (and SoCs) provide ways of introducing
+an error for test and debug purposes. This feature enables tools like [MemTest86](https://www.memtest86.com/)
+to inject ECC errors when running tests - keep in mind that ECC injection is
+available only in paid versions of MemTest86 run from UEFI and needs to be
+enabled either from the menu or in the configuration file. The most important
```

Comment:
@krystian-hebel maybe You should also mention that You have built coreboot with tianocore payload to lauch the Memtest Pro.